Add `CesiumITwinClient` and content loaders

azrogers opened the pull request:
Including the contents of former PRs #1121 and #1123, this PR expands on the additions by adding a `CesiumITwinClient` library, providing functionality comparable to `CesiumIonClient` for the Bentley iTwin APIs. This functionality primarily revolves around the API calls necessary to obtain access and refresh tokens and list and request available 3D Tiles tilesets from their various sources (Cesium Curated Content, Mesh Export, and Reality Data).

kring commented on Cesium3DTilesSelection/src/IModelMeshExportContentLoader.cpp:
```diff
@@ -0,0 +1,222 @@
+    const std::string& iTwinAccessToken,
+    const CesiumGeospatial::Ellipsoid& ellipsoid) {
+  CesiumUtility::Uri getExportsUri("https://api.bentley.com/mesh-export/");
+  CesiumUtility::UriQuery getExportsQueryParams("");
```
`("")` isn't needed since you added a default constructor.

kring commented on Cesium3DTilesSelection/src/IModelMeshExportContentLoader.h:
```diff
@@ -0,0 +1,38 @@
+public:
+  IModelMeshExportContentLoader(
+      std::unique_ptr<TilesetContentLoader>&& pAggregatedLoader)
+      : _pAggregatedLoader(std::move(pAggregatedLoader)) {}
```
Move this implementation to the .cpp file.

kring commented on Cesium3DTilesSelection/src/ITwinCesiumCuratedContentLoader.h:
```diff
@@ -0,0 +1,49 @@
+      const std::shared_ptr<CesiumAsync::IAssetAccessor>& pAggregatedAccessor);
+
+private:
+  friend class ITwinCesiumCuratedContentAssetAccessor;
```
There's no such class.

kring commented on Cesium3DTilesSelection/src/ITwinRealityDataContentLoader.cpp:
```diff
@@ -0,0 +1,297 @@
+  }
+
+  const std::vector<CesiumAsync::IAssetAccessor::THeader> headers = {
+      {"Authorization", "Bearer " + iTwinAccessToken},
```
Do we need to implement token refresh for this and `IModelMeshExportContentLoader`?

azrogers replied:
That's a good point. Token refresh here is different than in the ion case - we do need to request a new link after an hour, but after an hour we'll also need to obtain a new access token, and handling the token refresh process is I think beyond the scope of what a tileset loader should do. So I think we'll need a callback to get a new access token, and with that access token we'll need to recreate the tileset loader.

kring commented on Cesium3DTilesSelection/src/ITwinUtilities.cpp:
```diff
@@ -0,0 +1,75 @@
+    const CesiumAsync::IAssetResponse* pResponse,
+    CesiumUtility::ErrorList& errors) {
+  rapidjson::Document jsonResponse;
+  const std::span<const std::byte> data = pResponse->data();
```
You should either check `pResponse` against nullptr, or pass it into this function by reference instead of pointer.

kring commented on Cesium3DTilesSelection/src/Tileset.cpp:
```diff
@@ -1,3 +1,4 @@
+#include "Cesium3DTilesSelection/TilesetLoaderFactory.h"
```
Angle brackets.

kring commented on Cesium3DTilesSelection/src/TilesetContentManager.h:
```diff
@@ -1,12 +1,13 @@
 #pragma once
 
+#include "Cesium3DTilesSelection/TilesetLoaderFactory.h"
```
Angle brackets.

kring commented on Cesium3DTilesSelection/src/TilesetContentManager.h:
```diff
@@ -57,13 +58,27 @@ class TilesetContentManager
       int64_t ionAssetID,
       const std::string& ionAccessToken,
-      const std::string& ionAssetEndpointUrl = "https://api.cesium.com/");
+      const std::string& ionAssetEndpointUrl = "https://api.cesium.com/")
```
Implementation should go in the .cpp file.

kring reviewed (commented):
This is a big PR and I have lots of comments, but overall this looks great!

azrogers commented on CesiumITwinClient/include/CesiumITwinClient/Resources.h:
```diff
@@ -0,0 +1,456 @@
+  /** @brief Indicates the state of the iModel. */
+  IModelState state;
+  /**
+   * @brief The maximum rectangular area on the Earth which encloses the iModel.
```
The parameter descriptions here were grabbed directly from the docs, and that's the wording they use: https://developer.bentley.com/apis/imodels-v2/operations/get-imodel-details/#imodel

azrogers commented:
For my own notes, three things left in this review:
- Removing the `listAllAvailableResources` method and associated classes.
- Implementing token refresh in the iModel Mesh Export and iTwin Reality Data content loaders, as mentioned above.
- Testing everything out and resolving CI issues.

azrogers commented:
Removed the `listAllAvailableResources` method and associated classes. I implemented token refresh for the reality data loader (the user provides a callback that returns a future that resolves to a new access token), but it turns out the blob storage URL returned by the mesh export API remains valid for an indefinite period, and doesn't require an access token. Or, at the very least, it doesn't expire within 24 hours, since the link I obtained yesterday still works today. So I believe that one won't need any token refresh functionality.

kring commented on Cesium3DTilesSelection/src/ITwinCesiumCuratedContentLoader.h:
```diff
@@ -0,0 +1,29 @@
+
+class ITwinCesiumCuratedContentLoader : public CesiumIonTilesetLoader {
+public:
+  using AuthorizationHeaderChangeListener = std::function<
+      void(const std::string& header, const std::string& headerValue)>;
```
Remove this and used the one on `CesiumIonTilesetLoader`?

kring reviewed (commented):
A few more small things. This is really close now!

azrogers commented:
@kring I don't know what's going on with vcpkg for clang-tidy CI, but I believe everything's fixed.

kring commented:
Thanks @azrogers!
